File: java/src/DisjointSets.java
public class DisjointSets {

	public static int[] createSets(int size) {
		int[] p = new int[size];
		for (int i = 0; i < size; i++)
			p[i] = i;
		return p;
	}

	public static int root(int[] p, int x) {
		return x == p[x] ? x : (p[x] = root(p, p[x]));
	}

	public static void unite(int[] p, int a, int b) {
		a = root(p, a);
		b = root(p, b);
		if (a != b)
			p[a] = b;
	}

	// Usage example
	public static void main(String[] args) {
		int[] p = createSets(10);
		System.out.println(false == (root(p, 0) == root(p, 9)));
		unite(p, 0, 9);
		System.out.println(true == (root(p, 0) == root(p, 9)));
	}
}

File: kotlin/DisjointSets.kt
object DisjointSets {

    fun root(p: IntArray, x: Int): Int {
        return if (x == p[x]) {
            x
        } else {
            p[x] = root(p, p[x])
            p[x]
        }
    }

    fun unite(p: IntArray, a: Int, b: Int): Boolean {
        val ra = root(p, a)
        val rb = root(p, b)
        p[ra] = rb
        return ra != rb
    }

    fun createSets(size: Int) = IntArray(size) { it }

    // Usage example
    @JvmStatic
    fun main(args: Array<String>) {
        val p = createSets(10)
        println(false == (root(p, 0) == root(p, 9)))
        unite(p, 0, 9)
        println(true == (root(p, 0) == root(p, 9)))
    }
}
